File: email_service.py
import asyncio
import base64
import hashlib
import hmac
import logging
import os
import re
import smtplib
import uuid
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.utils import formatdate, make_msgid
from datetime import datetime
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

log = logging.getLogger("email_service")

UNSUBSCRIBE_SECRET = os.getenv("JWT_SECRET", "change-me-in-production")
FRONTEND_URL = os.getenv("FRONTEND_URL", "https://mail-hub.pro")


def generate_unsubscribe_token(email: str) -> str:
    email_b64 = base64.urlsafe_b64encode(email.encode()).decode()
    sig = hmac.new(UNSUBSCRIBE_SECRET.encode(), email.encode(), hashlib.sha256).hexdigest()
    return f"{email_b64}.{sig}"


def verify_unsubscribe_token(token: str) -> str | None:
    try:
        email_b64, sig = token.rsplit(".", 1)
        email = base64.urlsafe_b64decode(email_b64).decode()
        expected = hmac.new(UNSUBSCRIBE_SECRET.encode(), email.encode(), hashlib.sha256).hexdigest()
        if hmac.compare_digest(sig, expected):
            return email
    except Exception:
        pass
    return None


def build_unsubscribe_url(email: str) -> str:
    return f"{FRONTEND_URL}/unsubscribe/{generate_unsubscribe_token(email)}"


def _strip_html(html: str) -> str:
    text = re.sub(r"<br\s*/?>", "\n", html)
    text = re.sub(r"</p>", "\n\n", text)
    text = re.sub(r"<[^>]+>", "", text)
    text = re.sub(r"\n{3,}", "\n\n", text).strip()
    return text


def _send_single(
    host: str,
    port: int,
    user: str,
    password: str,
    from_email: str,
    from_name: str,
    to_email: str,
    subject: str,
    body: str,
    unsubscribe_url: str | None = None,
    signature: str | None = None,
) -> tuple[bool, str | None]:
    try:
        domain = from_email.split("@")[1] if "@" in from_email else "localhost"

        if signature:
            sig_html = (
                '<div style="margin-top:24px;padding-top:16px;border-top:1px solid #e5e7eb;">'
                f'{signature}'
                '</div>'
            )
            body = body + sig_html

        if unsubscribe_url:
            unsub_html = (
                '<div style="margin-top:30px;padding-top:15px;border-top:1px solid #e5e7eb;'
                'text-align:center;font-size:12px;color:#9ca3af;">'
                'If you no longer wish to receive these emails, '
                f'<a href="{unsubscribe_url}" style="color:#6366f1;">unsubscribe here</a>.'
                '</div>'
            )
            unsub_plain = f"\n\n---\nTo unsubscribe: {unsubscribe_url}"
            body = body + unsub_html
        else:
            unsub_plain = ""

        msg = MIMEMultipart("alternative")
        msg["From"] = f"{from_name} <{from_email}>" if from_name else from_email
        msg["To"] = to_email
        msg["Subject"] = subject
        msg["Date"] = formatdate(localtime=True)
        msg["Message-ID"] = make_msgid(domain=domain)
        msg["MIME-Version"] = "1.0"
        msg["X-Mailer"] = "MailHub/2.0"

        if unsubscribe_url:
            msg["List-Unsubscribe"] = f"<{unsubscribe_url}>"
            msg["List-Unsubscribe-Post"] = "List-Unsubscribe=One-Click"

        plain_text = _strip_html(body) + unsub_plain
        msg.attach(MIMEText(plain_text, "plain", "utf-8"))
        msg.attach(MIMEText(body, "html", "utf-8"))

        with smtplib.SMTP(host, port, timeout=15) as server:
            server.ehlo(domain)
            server.starttls()
            server.ehlo(domain)
            server.login(user, password)
            server.send_message(msg)
        return True, None
    except Exception as e:
        log.error("SMTP send failed to %s: %s", to_email, e)
        return False, str(e)


async def send_test_email(
    host: str,
    port: int,
    user: str,
    password: str,
    from_email: str,
    from_name: str,
    to_email: str,
) -> tuple[bool, str | None]:
    subject = "Ch-Scraper — SMTP Test"
    body = """
    <div style="font-family: Arial, sans-serif; max-width: 480px; margin: 0 auto; padding: 24px;">
      <h2 style="color: #4f46e5;">SMTP Connection Successful</h2>
      <p style="color: #374151; line-height: 1.6;">
        Your SMTP settings are working correctly. Emails from your Ch-Scraper account
        will be sent through this email address.
      </p>
      <hr style="border: none; border-top: 1px solid #e5e7eb; margin: 20px 0;" />
      <p style="color: #9ca3af; font-size: 12px;">
        This is an automated test from Ch-Scraper.
      </p>
    </div>
    """
    return await asyncio.to_thread(
        _send_single, host, port, user, password, from_email, from_name, to_email, subject, body
    )


async def queue_emails(db: AsyncSession, sent_records: list):
    from models import SentEmail, BatchEmail, Business, Company, UnsubscribedEmail

    if not sent_records:
        return

    company_id = sent_records[0].company_id
    comp_q = await db.execute(select(Company).where(Company.id == company_id))
    company = comp_q.scalar_one_or_none()

    if not company or not company.smtp_enabled or not company.smtp_host or not company.smtp_pass:
        for record in sent_records:
            record.status = "failed"
            record.sent_at = datetime.utcnow()
        await db.commit()
        log.warning("Company %s has no SMTP configured — marking %d emails as failed", company_id, len(sent_records))
        return

    unsub_q = await db.execute(select(UnsubscribedEmail.email))
    unsubscribed_set = {row[0].lower() for row in unsub_q.all()}

    async def process(record):
        be_q = await db.execute(
            select(Business.email)
            .join(BatchEmail, BatchEmail.business_id == Business.id)
            .where(BatchEmail.id == record.batch_email_id)
        )
        recipient = be_q.scalar()
        if not recipient:
            record.status = "failed"
        elif recipient.lower() in unsubscribed_set:
            record.status = "unsubscribed"
            log.info("Skipped unsubscribed email: %s", recipient)
        else:
            unsub_url = build_unsubscribe_url(recipient)
            success, error = await asyncio.to_thread(
                _send_single,
                company.smtp_host,
                company.smtp_port or 587,
                company.smtp_user,
                company.smtp_pass,
                company.smtp_from_email or company.smtp_user,
                company.smtp_from_name or company.name,
                recipient,
                record.subject,
                record.body,
                unsub_url,
                company.email_signature,
            )
            record.status = "sent" if success else "failed"
            if error:
                log.error("Failed to send to %s: %s", recipient, error)
        record.sent_at = datetime.utcnow()

    for record in sent_records:
        await process(record)

    await db.commit()

File: app/Services/EmailService.php
<?php
namespace App\Services;

use Illuminate\Support\Facades\Log;
use Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport;
use Symfony\Component\Mailer\Mailer as SymfonyMailer;
use Symfony\Component\Mime\Email;
use Symfony\Component\Mime\Address;
use Symfony\Component\Mime\Part\DataPart;

class EmailService
{
    private function secret(): string
    {
        return config('app.jwt_secret') ?? env('JWT_SECRET', 'change-me');
    }

    public function generateUnsubscribeToken(string $email): string
    {
        $b64 = str_replace(['+', '/'], ['-', '_'], base64_encode($email));
        $sig = hash_hmac('sha256', $email, $this->secret());
        return $b64 . '.' . $sig;
    }

    public function verifyUnsubscribeToken(string $token): ?string
    {
        $parts = explode('.', $token, 2);
        if (count($parts) !== 2) return null;
        [$b64, $sig] = $parts;
        $email = base64_decode(str_replace(['-', '_'], ['+', '/'], $b64));
        if ($email === false) return null;
        $expected = hash_hmac('sha256', $email, $this->secret());
        return hash_equals($expected, $sig) ? $email : null;
    }

    public function buildUnsubscribeUrl(string $email): string
    {
        $token = $this->generateUnsubscribeToken($email);
        return rtrim(config('app.frontend_url', 'http://localhost:5173'), '/') . '/unsubscribe/' . $token;
    }

    public function maskEmail(string $email): string
    {
        if (!str_contains($email, '@')) return '***';
        [$local, $domain] = explode('@', $email, 2);
        if (strlen($local) <= 2) {
            $masked = $local[0] . '***';
        } else {
            $masked = $local[0] . '***' . substr($local, -1);
        }
        return $masked . '@' . $domain;
    }

    public function embedInlineDataImages(Email $email, string $htmlBody): string
    {
        return preg_replace_callback(
            '/(<img\b[^>]*\bsrc=)(["\'])data:(image\/[a-zA-Z0-9.+-]+);base64,([^"\']+)\2/i',
            function (array $matches) use ($email) {
                $imageData = base64_decode($matches[4], true);
                if ($imageData === false) {
                    return $matches[0];
                }

                $extension = match (strtolower($matches[3])) {
                    'image/jpeg' => 'jpg',
                    'image/svg+xml' => 'svg',
                    default => strtolower(substr($matches[3], strlen('image/'))),
                };
                $contentId = 'inline-image-' . count($email->getAttachments()) . '@professionalclean.local';
                $part = (new DataPart($imageData, "signature-logo.{$extension}", $matches[3]))
                    ->setContentId($contentId)
                    ->asInline();
                $email->addPart($part);

                return $matches[1] . $matches[2] . 'cid:' . $contentId . $matches[2];
            },
            $htmlBody,
        );
    }

    private function normalizeHtmlForEmail(string $html): string
    {
        return preg_replace_callback(
            '/<p(\b[^>]*)>/i',
            function (array $m) {
                $attrs = $m[1];
                if (preg_match('/\bstyle=["\']([^"\']*)["\']/', $attrs, $s)) {
                    $style = rtrim($s[1], ';') . ';margin:0 0 14px 0;padding:0;';
                    $attrs = preg_replace('/\bstyle=["\'][^"\']*["\']/', "style=\"{$style}\"", $attrs);
                } else {
                    $attrs .= ' style="margin:0 0 14px 0;padding:0;"';
                }
                return '<p' . $attrs . '>';
            },
            $html,
        );
    }

    public function sendSingle(
        string $host, int $port, string $user, string $pass,
        string $fromEmail, string $fromName, string $toEmail,
        string $subject, string $htmlBody,
        ?string $unsubUrl = null, ?string $signature = null
    ): array {
        try {
            $htmlBody = $this->normalizeHtmlForEmail($htmlBody);

            if ($signature) {
                $htmlBody .= '<div style="margin-top:24px;padding-top:16px;border-top:1px solid #e5e7eb;">'
                    . $signature . '</div>';
            }

            $plainText = strip_tags(str_replace(
                ['<br>', '<br/>', '<br />', '</p>'],
                ["\n", "\n", "\n", "\n\n"],
                $htmlBody
            ));

            if ($unsubUrl) {
                $htmlBody .= '<div style="margin-top:30px;padding-top:15px;border-top:1px solid #e5e7eb;'
                    . 'text-align:center;font-size:12px;color:#9ca3af;">If you no longer wish to receive '
                    . 'these emails, <a href="' . $unsubUrl . '" style="color:#6366f1;">unsubscribe here</a>.</div>';
                $plainText .= "\n\n---\nTo unsubscribe: " . $unsubUrl;
            }

            $transport = new EsmtpTransport($host, $port);
            $transport->setUsername($user);
            $transport->setPassword($pass);

            $email = (new Email())
                ->from(new Address($fromEmail, $fromName))
                ->replyTo($fromEmail)
                ->to($toEmail)
                ->subject($subject)
                ->text($plainText);

            $htmlBody = $this->embedInlineDataImages($email, $htmlBody);
            $email->html($htmlBody);

            if ($unsubUrl) {
                $email->getHeaders()->addTextHeader('List-Unsubscribe', '<' . $unsubUrl . '>');
                $email->getHeaders()->addTextHeader('List-Unsubscribe-Post', 'List-Unsubscribe=One-Click');
            }

            (new SymfonyMailer($transport))->send($email);
            return [true, null];
        } catch (\Exception $e) {
            Log::warning('SMTP send failed', [
                'host' => $host,
                'port' => $port,
                'to_email' => $toEmail,
                'error' => $e->getMessage(),
            ]);
            return [false, $e->getMessage()];
        }
    }

    public function checkSmtpAuth(string $host, int $port, string $user, string $pass): array
    {
        try {
            $isDirectSsl = $port === 465;
            $address = ($isDirectSsl ? 'ssl://' : 'tcp://') . $host . ':' . $port;
            $context = stream_context_create(['ssl' => ['verify_peer' => false, 'verify_peer_name' => false]]);

            $fp = @stream_socket_client($address, $errno, $errstr, 15, STREAM_CLIENT_CONNECT, $context);
            if (!$fp) {
                return [false, "Cannot connect to {$host}:{$port} — {$errstr}"];
            }
            stream_set_timeout($fp, 10);

            $read = function () use ($fp): array {
                $code = 0; $text = '';
                while ($line = fgets($fp, 1024)) {
                    $code = (int)substr($line, 0, 3);
                    $text .= trim(substr($line, 4)) . ' ';
                    if (isset($line[3]) && $line[3] === ' ') break;
                }
                return [$code, trim($text)];
            };
            $send = fn(string $cmd) => fwrite($fp, $cmd . "\r\n");

            [$code] = $read();
            if ($code !== 220) { fclose($fp); return [false, "Unexpected greeting (code {$code})"]; }

            $send("EHLO localhost");
            [$code, $ehloText] = $read();
            if ($code !== 250) { fclose($fp); return [false, "EHLO failed (code {$code})"]; }

            if (!$isDirectSsl && stripos($ehloText, 'STARTTLS') !== false) {
                $send("STARTTLS");
                [$code] = $read();
                if ($code !== 220) { fclose($fp); return [false, "STARTTLS rejected (code {$code})"]; }
                stream_socket_enable_crypto($fp, true, STREAM_CRYPTO_METHOD_TLS_CLIENT);
                $send("EHLO localhost");
                $read();
            }

            $send("AUTH LOGIN");
            [$code] = $read();
            if ($code === 334) {
                $send(base64_encode($user));
                $read();
                $send(base64_encode($pass));
                [$code, $msg] = $read();
            } else {
                $send("AUTH PLAIN " . base64_encode("\0{$user}\0{$pass}"));
                [$code, $msg] = $read();
            }

            $send("QUIT");
            fclose($fp);

            if ($code === 235) return [true, null];
            return [false, "Authentication failed ({$code}): {$msg}"];
        } catch (\Throwable $e) {
            return [false, $e->getMessage()];
        }
    }

    public function sendTestEmail(
        string $host, int $port, string $user, string $pass,
        string $fromEmail, string $fromName, string $toEmail
    ): array {
        $body = '<div style="font-family:Arial,sans-serif;max-width:480px;margin:0 auto;padding:24px;">'
            . '<h2 style="color:#4f46e5;">SMTP Connection Successful</h2>'
            . '<p style="color:#374151;line-height:1.6;">Your SMTP settings are working correctly.</p>'
            . '</div>';
        return $this->sendSingle($host, $port, $user, $pass, $fromEmail, $fromName, $toEmail, 'ProfessionalClean — SMTP Test', $body);
    }

}
